File: vercel-ai/tour-of-agents/src/workflow-sequential.ts
import * as restate from "@restatedev/restate-sdk";
import { openai } from "@ai-sdk/openai";
import {generateText, Output, wrapLanguageModel} from "ai";
import { durableCalls } from "@restatedev/vercel-ai-middleware";
import { convertCurrency, processPayment } from "./utils/utils";
import {ClaimData, ClaimPromptSchema} from "./utils/types";
import { Context } from "@restatedev/restate-sdk";
import { z } from "zod";
const schema = restate.serde.schema;

// <start_here>
const process = async (ctx: Context, {prompt}: {prompt: string}) => {
  const model = wrapLanguageModel({
    model: openai("gpt-4o"),
    middleware: durableCalls(ctx, { maxRetryAttempts: 3 }),
  });

  // Step 1: Parse the claim document (LLM step)
  const { output } = await generateText({
    model,
    system:
      "Extract the claim amount, currency, category, and description.",
    prompt,
    output: Output.object({schema: ClaimData})
  });

  // Step 2: Analyze the claim (LLM step)
  const { output: valid } = await generateText({
    model,
    system:
      "You are a claims analyst. Assess whether this claim is valid and determine the approved amount.",
    prompt: `Claim: ${JSON.stringify(output)}`,
    output: Output.object({schema: z.object({valid: z.boolean()})}),
  });

  if (!valid) {
    return { analysis: "Claim is invalid", amountUsd: 0, confirmation: false };
  }

  // Step 3: Convert currency (regular step)
  const amountUsd = await ctx.run("Convert currency", async () =>
    convertCurrency(output.amount, output.currency, "USD"),
  );

  // Step 4: Process reimbursement (regular step)
  const confirmation = await ctx.run("Process payment", async () =>
    processPayment(ctx.rand.uuidv4(), amountUsd),
  );

  return { analysis: "Claim is valid", amountUsd, confirmation };
};
// <end_here>

const agent = restate.service({
  name: "ClaimReimbursement",
  handlers: {
    process: restate.createServiceHandler(
      { input: schema(ClaimPromptSchema) },
      process,
    ),
  },
});

restate.serve({ services: [agent] });

File: langchain-python/tour-of-agents/app/workflow_sequential.py
"""Sequential pipeline: parse → analyze → currency conversion → payment.
Each step is durable, so the workflow can resume mid-pipeline after a
failure without re-running already-completed steps."""

import restate
from langchain.agents import create_agent
from langchain.chat_models import init_chat_model

from restate.ext.langchain import RestateMiddleware

from utils.models import ClaimData, ClaimPrompt
from utils.utils import convert_currency, process_payment


# <start_here>
claim_service = restate.Service("ClaimReimbursement")


@claim_service.handler()
async def process(ctx: restate.Context, req: ClaimPrompt) -> dict:
    # Step 1: Parse the claim document (structured-output LLM step).
    parse_agent = create_agent(
        model=init_chat_model("openai:gpt-4o-mini"),
        tools=[],
        system_prompt="Extract the claim amount, currency, category, and description.",
        response_format=ClaimData,
        middleware=[RestateMiddleware()],
    )
    parsed = await parse_agent.ainvoke(
        {"messages": [{"role": "user", "content": req.message}]}
    )
    claim: ClaimData = ClaimData.model_validate(parsed["structured_response"])

    # Step 2: Analyze the claim (LLM step).
    analysis_agent = create_agent(
        model=init_chat_model("openai:gpt-4o-mini"),
        tools=[],
        system_prompt="Assess whether this claim is valid and determine the approved amount.",
        middleware=[RestateMiddleware()],
    )
    analysis = await analysis_agent.ainvoke(
        {"messages": [{"role": "user", "content": f"Claim: {claim.model_dump_json()}"}]}
    )

    # Step 3: Convert currency (regular durable step).
    amount_usd = await ctx.run_typed(
        "Convert currency",
        convert_currency,
        amount=claim.amount,
        source=claim.currency,
        target="USD",
    )

    # Step 4: Process reimbursement (regular durable step).
    confirmation = await ctx.run_typed(
        "Process payment",
        process_payment,
        claim_id=str(ctx.uuid()),
        amount=amount_usd,
    )

    return {
        "analysis": analysis["messages"][-1].content,
        "amount_usd": amount_usd,
        "confirmation": confirmation,
    }


# <end_here>


if __name__ == "__main__":
    import asyncio

    import hypercorn
    import hypercorn.asyncio

    app = restate.app(services=[claim_service])
    conf = hypercorn.Config()
    conf.bind = ["0.0.0.0:9080"]
    asyncio.run(hypercorn.asyncio.serve(app, conf))
